Fix a mis-parse of NPY v2.0 and 3.0 headers

nmcclatchey opened the pull request:
CNPY originally only supported NPY 1.0 headers, with a 2-byte length field. This PR causes `parse_npy_header()` to respond to NPY 2.0 and 3.0 by reading the header length as 4 bytes.

nmcclatchey committed:
Fix a mis-parse of NPY v2.0 and 3.0 headers

CNPY originally only supported NPY 1.0 headers, with a 2-byte length
field. This commit causes parse_npy_header to respond to NPY 2.0 and 3.0
by reading the header length as 4 bytes.

ZJUGuoShuai commented:
I only saw version 2.0 from [NEP 1](https://numpy.org/neps/nep-0001-npy-format.html#format-specification-version-2-0), can you tell me where is version 3.0 defined? Thanks.

nmcclatchey commented:
You can find all 3 versions defined on numpy.org. Specifically, on [the numpy.lib.format page](https://numpy.org/devdocs/reference/generated/numpy.lib.format.html).

Version 3.0 merely swaps ASCII strings to Unicode strings.

s-trinh commented on cnpy.cpp:
```diff
@@ -63,8 +63,13 @@ void cnpy::parse_npy_header(unsigned char* buffer,size_t& word_size, std::vector
+      header_len = *reinterpret_cast<uint32_t*>(buffer+8);
+    else
+      header_len = *reinterpret_cast<uint16_t*>(buffer+8);
+    std::string header(reinterpret_cast<char*>(buffer+(extended_header ? 11 : 9)),header_len);
```
Since `header_len` is `uint16_t`, I think it should be?

```suggestion
    std::string header(reinterpret_cast<char*>(buffer+(extended_header ? 12 : 10)),header_len);
```

nmcclatchey replied:
I think you're right. It's been working because the only use of `header` has been based on relative locations discovered using `find` and similar, but we could improve performance (ever so slightly) by using the correct offset.

nmcclatchey replied:
Fixed. Thanks for pointing this out.

nmcclatchey committed:
Bugfix: `header` string was oversized

The `header` string was started too early, and had included the final byte of the length field (and omitting the final byte). This commit adjusts where it starts to fix the issue.

Note: No ill effects occurred earlier because each parsed field is located using `find`, and the final byte is typically `}` and irrelevant to the parsed fields.
